### PR TITLE
Updated build pipeline for distribution of compiled assets

### PR DESCRIPTION
**Pull request by kmalakoff**
I've updated Clipboard.js to use webpack for pre-compiling assets for the browser to address:

https://github.com/zenorocha/clipboard.js/issues/27

I've commented where you can make the modules separate for the delegate and emitter modules (if you want, but I'm assuming bundling is preferred): https://github.com/kmalakoff/clipboard.js/blob/master/webpack.config.js#L15

### CODE REVIEW

**Comment by brancusi** [comment]
:thumbsup:

File: gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+var gulp = require('gulp');
```
Comment (heldr) [inline gulpfile.js]: uhm...Is gulp and browsersync really necessary?(both tools are nice, but I believe it's overkill for a simple library)

So, why don't use webpack over npm scripts https://github.com/zenorocha/clipboard.js/blob/master/package.json#L43-L49 instead?

File: package.json [hunk 2]
```diff
@@ -4,30 +4,29 @@
...
-    "babelify": "^6.3.0",
-    "browserify": "^11.2.0",
     "delegate-events": "^1.1.1",
     "tiny-emitter": "^1.0.0"
   },
   "devDependencies": {
+    "babel-core": "^5.8.23",
+    "babel-loader": "^5.3.2",
+    "babel-runtime": "^5.8.20",
+    "babelify": "^6.3.0",
```
Comment (heldr) [inline package.json]: Why leave babelify and browserify if you're using webpack?

Comment (heldr) [inline package.json]: Now I get it! You're using it for test proposes.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import Clipboard from './src/clipboard';
+import ClipboardAction from './src/clipboard-action';
+import DelegateEvents from 'delegate-events';
+import TinyEmitter from 'tiny-emitter';
+
+// expose bundled libraries
+Clipboard.ClipboardAction = ClipboardAction;
+Clipboard.DelegateEvents = DelegateEvents;
+Clipboard.TinyEmitter = TinyEmitter;
```
Comment (heldr) [inline index.js]: I personally wouldn't expose internal deps that much on a entrypoint. Local scope brings more maintainability.

Which  means:
```js
import Clipboard from './src/clipboard';
```
only and internal deps whenever it's required on a specific file.

File: webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    libraryTarget: 'umd',
+    filename: 'clipboard.js'
+  },
+  module: {
+    loaders: [{test: /\.js?$/, exclude: /(node_modules|bower_components)/, loader: 'babel?optional[]=runtime&stage=0'}]
+  },
+  externals: [
+    // delegate-events and tiny-emitter could go here if you want them externally provided
+    // {library: {root: 'LibraryWindowSymbol', amd: 'library', commonjs: 'library', commonjs2: 'library'}},
+  ]
```
Comment (zenorocha) [inline webpack.config.js]: No need to expose them

File: gulpfile.js (hunk 1 above)
Comment (zenorocha) [inline gulpfile.js]: +1 for npm scripts only

**Comment by kmalakoff** [comment]
Hey guys. What's the call on merging this or an adapted version?

**Comment by zenorocha** [comment]
@kmalakoff have you seen the inline comments?

**Comment by kmalakoff** [comment]
@zenorocha are you talking about the comments in this pull request? If so, it's up to you guys if you want to tweak my submission - it works perfectly well as is.

File: index.js (hunk 3 above)
Comment (kmalakoff) [inline index.js]: If you are bundling libraries, it is common practice to expose them so they can be accessed by library users.

File: webpack.config.js (hunk 4 above)
Comment (kmalakoff) [inline webpack.config.js]: If you are bundling libraries, it is common practice to expose them so they can be accessed by library users.

File: gulpfile.js (hunk 1 above)
Comment (kmalakoff) [inline gulpfile.js]: I think gulp is easier to understand and maintain. Your choice if you want to reimplement my work.

**Comment by kmalakoff** [comment]
@zenorocha I've responded to comments. 

Basically, I'm not onboard with the proposed changes (well, removing browsersync is fine) and have devoted enough time to this. What I'm looking for is a compiled version of your library that does not require babel so anything that fits that need is OK by me.

The ball is in your court.

File: gulpfile.js (hunk 1 above)
Comment (kmalakoff) [inline gulpfile.js]: Agreed that browsersync is not necessary and can be removed. It was just from another project that I used as reference.

**Comment by kmalakoff** [comment]
How's the review coming? Should be a simple merge and you can add a task to reimplement later.

**Comment by zenorocha** [comment]
This PR is filled with personal preferences and opinionated tooling that are not truly necessary. 

Here are two different ways of using Clipboard.js with Webpack:
* [Compiling Clipboard.js' source using Webpack](https://gist.github.com/zenorocha/88db55b251af6f3df41c)
* [Bundling Clipboard.js' distribution with Webpack](https://gist.github.com/zenorocha/53daffe28de983907579)

**Comment by kmalakoff** [comment]
I love this comment in https://github.com/zenorocha/clipboard.js/issues/27:  `Your intentions were good, but your pull request not, period.`. 

You know I've submitted a similar improvement to other libraries and this was the first time it was rejected. 

I'm not saying it was perfect, but...@zenorocha thank you for the delays, runaround and the drama. :thumbsdown:

(FYI: put downs are not a good way to encourage contributions to your library)

**Comment by zenorocha** [comment]
There's absolutely no reason for this discussion to continue.

The real problem was tracked and resolved https://github.com/zenorocha/clipboard.js/commit/4b73122f810a0e52a7eacf47e8d3cfffa05584dc, not only for Webpack (which was your use case) but also for Browserify and other module loaders that relies on npm.

If you're want to post inflammatory arguments, you're in the wrong place and talking to the wrong person. I have zero interest in this kind of discussion.

**Pull request closed without merging**